geode-core/src/test/java/org/apache/geode/internal/cache/tier/sockets/ServerConnectionTest.java

```diff
@@ -139,4 +158,108 @@ public void post65NonSecureShouldThrow() {
         .hasMessage(HandShake_NO_SECURITY_CREDENTIALS_ARE_PROVIDED.getRawText());
   }
 
+  class TestMessage extends Message {
+    private final Lock lock = new ReentrantLock();
+    private final Condition testGate = lock.newCondition();
+    private boolean signalled = false;
+
+    public TestMessage() {
+      super(3, Version.CURRENT);
+      messageType = MessageType.REQUEST;
+      securePart = new Part();
+    }
+
+    @Override
+    public void recv() throws IOException {
+      try {
+        lock.lock();
+        testGate.await(10, TimeUnit.SECONDS);
+      } catch (InterruptedException e) {
+        throw new RuntimeException(e);
+      } finally {
+        lock.unlock();
+        if (!signalled) {
+          fail("Message never received continueProcessing call");
+        }
+      }
+    }
+
+    public void continueProcessing() {
+      lock.lock();
+      testGate.signal();
+      signalled = true;
+      lock.unlock();
+    }
+  }
+
+  class TestServerConnection extends LegacyServerConnection {
+
+    private TestMessage testMessage;
+
+    /**
+     * Creates a new <code>ServerConnection</code> that processes messages received from an edge
+     * client over a given <code>Socket</code>.
+     */
+    public TestServerConnection(Socket socket, InternalCache internalCache,
+        CachedRegionHelper helper, CacheServerStats stats, int hsTimeout, int socketBufferSize,
+        String communicationModeStr, byte communicationMode, Acceptor acceptor,
+        SecurityService securityService) {
+      super(socket, internalCache, helper, stats, hsTimeout, socketBufferSize, communicationModeStr,
+          communicationMode, acceptor, securityService);
+
+      setClientDisconnectCleanly(); // Not clear where this is supposed to be set in the timeout
+                                    // path
+    }
+
+    @Override
+    protected void doHandshake() {
+      ClientProxyMembershipID proxyID = mock(ClientProxyMembershipID.class);
+      when(proxyID.getDistributedMember()).thenReturn(mock(InternalDistributedMember.class));
+      HandShake handShake = mock(HandShake.class);
+      when(handShake.getMembership()).thenReturn(proxyID);
+      when(handShake.getVersion()).thenReturn(Version.CURRENT);
+
+      setHandshake(handShake);
+      setProxyId(proxyID);
+
+      processHandShake();
+      initializeCommands();
+
+      setFakeRequest();
+
+      long fakeId = -1;
+      MessageIdExtractor extractor = mock(MessageIdExtractor.class);
+      when(extractor.getUniqueIdFromMessage(getRequestMessage(), handShake,
+          Connection.DEFAULT_CONNECTION_ID)).thenReturn(fakeId);
+      setMessageIdExtractor(extractor);
+    }
+
+    @Override
+    void handleTermination(boolean timedOut) {
+      super.handleTermination(timedOut);
+      testMessage.continueProcessing();
+    }
+
+    private void setFakeRequest() {
+      testMessage = new TestMessage();
+      setRequestMsg(testMessage);
+    }
+  }
+
+  @Test
```

Comment:
This test is complicated enough that it could use a javadoc describing what it's testing and how it's going about it.

Reply:
Added.